### PR TITLE
Persist memory support in TextSet

### PR DESCRIPTION
**Pull request by qiuxin2012**
Persist memory support in TextSet
User could set memoryType to `PMEM` to cache train dataset to Persist Memory.

### CODE REVIEW

File: zoo/src/test/scala/com/intel/analytics/zoo/feature/pmem/PersistentMemorySpec.scala [hunk 1]
```diff
@@ -126,4 +128,19 @@ class PersistentMemorySpec extends ZooSpecHelper {
     assert(data.count() == 3)
     data.collect()
   }
+
+  "getting data in FeatureSet" should "be right" in {
```
Comment (zhichao-li) [inline zoo/src/test/scala/com/intel/analytics/zoo/feature/pmem/PersistentMemorySpec.scala]: add this test in `NativeSpec` as well?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/QARanker.scala [hunk 2]
```diff
@@ -34,7 +35,7 @@ case class QARankerParams(
     questionLength: Int = 10, answerLength: Int = 40,
     partitionNum: Int = 4, batchSize: Int = 200,
     nbEpoch: Int = 30, learningRate: Double = 0.001,
-    model: Option[String] = None)
+    model: Option[String] = None, memoryType: String = "DRAM")
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/QARanker.scala]: Add some doc in README about memoryType and what each available type does

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/FeatureSet.scala [hunk 3]
```diff
@@ -16,12 +16,14 @@
...
 
-import com.intel.analytics.bigdl.dataset.ByteRecord
+import com.intel.analytics.bigdl.dataset.{ByteRecord, Sample}
+import com.intel.analytics.bigdl.transform.vision.image.ImageFeature
 import com.intel.analytics.zoo.feature.DistributedFeatureSet
 import com.intel.analytics.zoo.feature.common.ArrayLike
 import org.apache.spark.rdd.RDD
 
 import scala.reflect.ClassTag
+import scala.reflect.classTag
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/FeatureSet.scala]: why import twice?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/FeatureSet.scala [hunk 4]
```diff
@@ -16,12 +16,14 @@
 
 package com.intel.analytics.zoo.feature.pmem
 
-import com.intel.analytics.bigdl.dataset.ByteRecord
+import com.intel.analytics.bigdl.dataset.{ByteRecord, Sample}
+import com.intel.analytics.bigdl.transform.vision.image.ImageFeature
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/pmem/FeatureSet.scala]: is this import used?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 5]
```diff
@@ -35,6 +35,8 @@ import scala.io.Source
 import java.io.PrintWriter
 
 import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.zoo.feature.FeatureSet
+import com.intel.analytics.zoo.feature.pmem.{DIRECT, DRAM, MemoryType, PMEM}
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: why need DIRECT PMEM?

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala [hunk 6]
```diff
@@ -563,11 +567,12 @@ class DistributedTextSet(var rdd: RDD[TextFeature]) extends TextSet {
...
 
   override def toDataSet: DataSet[Sample[Float]] = {
-    DataSet.rdd(rdd.map(_[Sample[Float]](TextFeature.sample)))
+    FeatureSet.rdd(rdd.map(_[Sample[Float]](TextFeature.sample)),
+      memoryType)
   }
 
   override def randomSplit(weights: Array[Double]): Array[TextSet] = {
-    rdd.randomSplit(weights).map(TextSet.rdd)
+    rdd.randomSplit(weights).map(v => TextSet.rdd(v))
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/text/TextSet.scala]: here why not involve memoryType?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 7]
```diff
@@ -373,7 +374,10 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
       nbEpoch: Int,
       validationData: TextSet)(implicit ev: TensorNumeric[T]): Unit = {
     KerasUtils.validateBatchSize(batchSize)
-    this.fit(toDataSet(x, batchSize), nbEpoch, toDataSet(validationData, batchSize))
+    val dataset = x.toDataSet
+    this.fit((dataset -> SampleToMiniBatch[Float](batchSize)).asInstanceOf[DataSet[MiniBatch[T]]],
+      nbEpoch, toDataSet(validationData, batchSize))
+    dataset.toDistributed().unpersist()
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: what about LocalTextSet? Here fit should both accept local and distributed

**Comment by hkvision** [comment]
Also, what about TextClassification example?

File: zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md [hunk 8]
```diff
@@ -55,6 +55,8 @@ See [here](#options) for more configurable options for this example.
 * `-e` `--nbEpoch` The number of iterations to train the model. Default is 30.
 * `-l` `--learningRate` The learning rate for the model. Default is 0.001.
 * `-m` `--model` Specify this option only if you want to load an existing KNRM model and in this case its path should be provided.
+* `--memoryType` optional. The default value is `DRAM`, you can change it to `PMEM` if have Intel Optane DC Persistent Memory.
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/examples/qaranker/README.md]: Change to `Memory type used for caching training data. Default is 'DRAM'. You can change it to 'PMEM' if you have Intel Optane DC Persistent Memory`

**Comment by qiuxin2012** [comment]
jenkins passed

**Pull request merged**